File: Problem Set 1/code/src/p03d_poisson.py
import numpy as np
import util

from linear_model import LinearModel


def main(lr, train_path, eval_path, pred_path):
    """Problem 3(d): Poisson regression with gradient ascent.

    Args:
        lr: Learning rate for gradient ascent.
        train_path: Path to CSV file containing dataset for training.
        eval_path: Path to CSV file containing dataset for evaluation.
        pred_path: Path to save predictions.
    """
    # Load training set
    x_train, y_train = util.load_dataset(train_path, add_intercept=False)
    x_eval, y_eval = util.load_dataset(eval_path, add_intercept=False)
    # *** START CODE HERE ***
    pois_reg = PoissonRegression(step_size=lr)
    pois_reg.fit(x_train, y_train)

    y_pred = pois_reg.predict(x_eval)

    # Plot results compared to the true result
    fig, ax = util.plt.subplots()
    # Index of the points that are error is less than 10%
    qf_idx = (np.linalg.norm(y_pred - y_eval, ord=1) / y_eval) < 0.1
    ax.plot(y_eval[qf_idx], y_pred[qf_idx], 'bx')
    ax.plot(y_eval[~qf_idx], y_pred[~qf_idx], 'rx')
    np.savetxt(pred_path, y_pred, fmt='%d')
    ax.set_xlabel('true counts')
    ax.set_ylabel('predict counts')
    fig.savefig('output/p03d.png')
    # *** END CODE HERE ***


class PoissonRegression(LinearModel):
    def fit(self, x, y):
        """Run gradient ascent to maximize likelihood for Poisson regression.

        Args:
            x: Training example inputs. Shape (m, n).
            y: Training example labels. Shape (m,).
        """
        # *** START CODE HERE ***
        # initiate theta with zeros

        m, n = x.shape
        self.theta = np.zeros(n)

        while True:
            # pick up a random index to perform mini batch gradient ascent with size of 100
            batch_size = 100
            idx_rnd = np.random.choice(range(m), size=batch_size)
            xi = x[idx_rnd]
            yi = y[idx_rnd]
            gradient = xi.T.dot(yi - np.exp(xi.dot(self.theta))) / batch_size

            self.theta += self.step_size * gradient
            if np.linalg.norm(self.step_size * gradient, ord=1) < self.eps: break
        # *** END CODE HERE ***

    def predict(self, x):
        """Make a prediction given inputs x.

        Args:
            x: Inputs of shape (m, n).

        Returns:
            Floating-point prediction for each input, shape (m,).
        """
        # *** START CODE HERE ***
        return np.exp(x.dot(self.theta))
        # *** END CODE HERE ***

File: Problem Set 1/Code/src_cpp/src_cpp/p03d_poisson.cpp
/*
source file to apply functions and class in the header file
*/

#include "p03d_poisson.hpp"
#include "util.hpp"

// constructors
PoissonRegression::PoissonRegression() : LinearModel() {};
PoissonRegression::PoissonRegression(const PoissonRegression& source) : LinearModel(source) {};
PoissonRegression::PoissonRegression(mat& theta, const double& step_size, const int& max_iter, const double& eps) : LinearModel(theta, step_size, max_iter, eps) {};
PoissonRegression::~PoissonRegression() {};

// Assignment operator
PoissonRegression& PoissonRegression::operator=(const PoissonRegression& source) {
	if (this == &source) {
		cout << "self-assignment check" << endl;
	}
	else { LinearModel::operator=(source);}
	return *this;
}

// functions
const void PoissonRegression::fit(const mat& x, const mat& y) {
	/*Run gradient ascent to maximize likelihood for Poisson regression.
	Args:
	x: Training example inputs.Shape(m, n).
	y : Training example labels.Shape(m, 1).*/

	// initiate theta with zeros
	uword m = x.n_rows, n = x.n_cols, batch_size=100;
	theta = arma::zeros(n, 1);

	while (true) {
		// randomly pick up indices
		uvec idx_rnd = randi<uvec>(batch_size, 1, distr_param(0, m - 1));
		mat xi = x.rows(idx_rnd);
		mat yi = y.rows(idx_rnd);
		mat gradient = xi.t() * (yi - exp(xi * theta)) / batch_size;

		theta += step_size * gradient;
		if (arma::norm(step_size * gradient, 1) < eps) { break; }
	}
}

const mat PoissonRegression::predict(const mat& x, const bool& p) {
	/*
	Make a prediction given inputs x.
	Args:
	x: Inputs of shape(m, n).
	Returns :
	Floating - point prediction for each input, shape(m, 1).*/
	return exp(x * theta);
}

void p03d_poisson(string dataset) {
	// loadata set
	std::string root, path_train, path_eval, savedr;
	root = "C:\\Users\\YB\\Documents\\GitHub\\Standford_CS229\\Problem Set 1\\Code\\data\\";
	savedr = "C:\\Users\\YB\\Documents\\GitHub\\Standford_CS229\\Problem Set 1\\Code\\src_cpp\\src_cpp\\output\\";
	path_train = dataset + "_train.csv";
	path_eval = dataset + "_valid.csv";

	tuple<mat, mat> data_train = util::load_dataset(root + path_train, 4, 4, false);
	tuple<mat, mat> data_eval = util::load_dataset(root + path_eval, 4, 4, false);

	// Get train/eval data
	mat X_train, y_train, X_eval, y_eval;

	X_train = get<0>(data_train);
	y_train = get<1>(data_train);

	X_eval = get<0>(data_eval);
	y_eval = get<1>(data_eval);

	// train dataset1, predict, save
	mat theta = arma::zeros(5);
	double step_size = 1e-7;
	unsigned int max_iter = 100;
	double eps = 1e-5;
	PoissonRegression pois_reg(theta, step_size, max_iter, eps);

	pois_reg.fit(X_train, y_train);

	mat y_pred;
	y_pred = pois_reg.predict(X_eval);
	y_pred.save(savedr + "p03d_pred_poisson_" + dataset + ".csv", arma::arma_ascii);
}